components/TextFields/src/MDCInputTextField.h

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <UIKit/UIKit.h>
+
+/**
+ A set of Contained Input View states outlined in the Material guidelines. These states overlap with
+ and extend UIControlState.
+ */
+typedef NS_OPTIONS(NSInteger, MDCContainedInputViewState) {
+  /**
+   The default state of the contained input view.
+   */
+  MDCContainedInputViewStateNormal = 1 << 0,
+  /**
+   The state the view is in during normal editing.
+   */
+  MDCContainedInputViewStateFocused = 1 << 1,
+  /**
+   This state most closely resembles the @c selected UIControlState.
+   */
+  MDCContainedInputViewStateActivated = 1 << 2,
+  /**
+   The error state.
+   */
+  MDCContainedInputViewStateErrored = 1 << 3,
```

Comment:
I don't think we should have an "errored" state. Having such a state (and responding to it) has been a source of maintenance issues and bugs in MDC Text Field. I think it's better to simplify the component and allow the application to determine when an error has occurred.  I think UIKit's UITextField APIs support this position:

*   Text field delegates allow the application to detect edits to the input text and respond. This is where an application would set an error message in any case. As a result, the application could easily manage any error-related theming with a single API call, `[textField applyErrorThemeWithScheme:containerScheme];`.
*   If the colors are being retained for each state, why not the border width, corner radii, fonts, underline thickness, etc? It seems inconsistent in general, but for error especially.

Reply:
In my experience with MDCTextField, the difficulties with the notion of an error state came from the fact that it cached helper text and error text and tied these various strings to different states. It became very messy and error prone to manage all that state. This approach _is_ a simplification. Toggling the error state just makes color changes. To change the text you call setText on an underline label.

The approach to error states in this textfield is not the same as MDCTextField.

Are you suggesting that the client manage multiple container schemes for different states? That is not how I thought theming was supposed to work.

In the theming extension the color schemes are adapted into color-centric view models that are retained and applied on a state by state basis. Regarding underline width and border width, those values are not derived from MDCColorSchemes. Things like that are the responsibility of the style objects. They have nothing to do with the theming extensions.

Reply:
+1 for keeping the error state.
It will make implementation simpler if the Error state is themed the same as other states. It will also make it easier to manage state combinations (like error + focused).

Reply:
> Are you suggesting that the client manage multiple container schemes for different states? 

Nope, I'm suggesting that TextField should have no internal concept of "error". Since determining an error is entirely up to the application and is not intrinsic to Text Fields, I would expect there be no API for an "error" state on Text Fields.  I'm not sure where you got the idea of different container schemes from, but I think if we need to style a text field for an error state, that a Theming Extension API provide that.  I tried to make it clear from my example code snippet, but perhaps I wasn't clear enough.

Reply:
+1 to @romoore's suggestion. I'm supportive of moving "error" to business logic via applying a different theming extension rather than encoding it as a distinct user interaction state. We'd still be able to accomplish the same intersections of stateful theming, but without the added overhead of a new state to manage internally to TextFields.

Reply:
If this is what you want @jverkoey I would like a chance to spike on it before I'm fully on board, is that alright?

Reply:
And @romoore, I see what you mean now. I read `[textField applyErrorThemeWithScheme:containerScheme];` as `[textField applyThemeWithScheme:containerScheme];`--I thought you were saying clients should just call the same method with different schemes to achieve error/non-error appearances.

Reply:
Spikes are @randallli's jurisdiction; make the case for what he's buying there.

Reply:
I'll reach out to him 👍

Reply:
+1 to moving error outside of the component. Button comes to mind as a good example of a component that does this externally. If I wanted an button to go to the error "state" I would call a dedicated theming API. (not that we have that API yet but that is how i would do it :))

Reply:
Ok I created [b/131170010](http://b/131170010) and made it a blocking issue. I put it as a 1 day spike. I also need to think about how this affects the other contained input views in disguise that I have planned, like InputChipView.

Reply:
@romoore or @jverkoey -- if the error state is achieved via the theming extension, and it's not a state per se, how should clients check if the text field is in an error state? Do something like `if (CGColorEqualToColor(textField.textColor.CGColor, myErrorColor.CGColor))`?

Reply:
I'd expect whether a text field is in an error state to be modeled in the model / business data, not the view.

Reply:
I'm not sure I follow :/ Is there like a line of pseudocode you could show me to help me better understand how a client should check to see if a text field is in an error state? Or are you saying they shouldn't be able to check like that?

Reply:
```swift
class ViewController {
  var isPhoneNumberValid = false

  func validateModel() {
    isPhoneNumberValid = checkPhoneNumber(textField.textLabel.text)

    if isPhoneNumberValid {
      textField.applyTheme(with: containerScheme)
    } else {
      textField.applyErrorTheme(with: containerScheme)
    }
  }
}

Reply:
Thanks! So you're saying they shouldn't be able to check if it's in an error state. I think that's a little strange tbh! I think if I were a client I would want to be able to check that kind of thing. Maybe I'm alone in that though haha.

Reply:
You can check that sort of thing; you'd just check it in the model / controller logic, rather than the view.

Reply:
Got it. This is just a big shift from the mindset I've been working under for months (the one that the guidelines seem to impart) where text fields have distinct error states.

Reply:
In all the examples we've used for MDC Text Fields, we never check if it's in an 'error state' - we just either set the error text to something (when there's a validation error) or nothing (when input passes without error). For these Text Fields, the user would be expected to apply theming along with any unique message they wish to add to the labels/placeholder/etc.   I think this could simplify both the component itself and the client app, since the client can retain full control of error management.  As it is, MDC Text Fields takes some of the error logic internally, leaving the client with only partial control.